### ISSUE TITLE
Wrong HTTP status code?

### ISSUE THREAD

**Issue by olalonde** [olalonde/follow-redirects#24]
Someone sent me this email and I referred them to this issue:

Hi James
I have been using "follow-redirects" for making ReST calls and its very good. Thanks for the wonderful package. I am however stuck up when the site does a redirect. In that case, something very strange happens. Instead of getting a 307, for some very weird reason I keep getting 400 response. 

Now I know this sounds very crazy as its the server which sends the response, and not the client, but whenever I use the normal "http" package, in my client code, instead of using the "follow-redirects", I get a 307 properly. Only for "follow-redirects" do I get a 400.

So is there any additional step that I need to do, to get this resolved.

**Comment by jamestalmage** [comment]
This is likely a duplicate of #12. 

When presented a 307 status code, (almost) all browsers will reuse the the original HTTP method when calling the secondary location. 


>*most browsers*
--> PUT Request from browser to original location
<-- 307 Response from server
--> **PUT** Request to secondary location
<-- ??? Response from secondary server

See [this project](brt) for a thorough suite of tests on how each browser handles redirects.

Currently, `follow-redirects` converts everything past the original request to use the `GET` method. So this is what is happening:

>*follow-redirects*
--> PUT Request from browser to original location
<-- 307 Response from server
--> **GET** Request to secondary location
<-- 4XX Response from secondary server because it is expecting a PUT


`follow-redirects` is kinda actually correct according to [the spec](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html).

> If the 307 status code is received in response to a request other than GET or HEAD, the user agent MUST NOT automatically redirect the request unless it can be confirmed by the user, since this might change the conditions under which the request was issued.

However, you can see from [the chart here](brt), that nearly all browsers ignore that and automatically follow the redirect. (*see note below*)

All that said, the intended **future** behavior of `follow-redirects` is to mimic the most common browser behavior (per #12). This will be part of the `1.0.0` release (and is the major blocking issue for that release). It will be a significant breaking change, so we will need to issue the appropriate warnings.

*Note:*

> Not important to this discussion, but interesting:
>
> Only Opera follows strict adherence to the spec. Opera actually posts a dialog allowing you to choose whether you follow the redirect or not, and whether it converts the request to a `GET` or resubmits with the original method (unfortunately, most users aren't going to understand the implications of that dialog, so while it strictly adheres, I don't see the benefit). 


[brt]: https://github.com/jamestalmage/browser-redirect-test

**Comment by jamestalmage** [comment]
An interim patch is to use the (undocumented) `makeRequest` callback.

```js
var http = require('follow-redirects').http;
var options = {
  host: 'www.example.com',
  path: '/my-resource',
  method: 'POST',
  makeRequest: function (options, cb, res) {
    //options - the original options passed in (to be mutated).
    //res - the result of the previous, redirected request (or null if this is the first request)
    if (res && res.statusCode !== 307) {
      options.method = 'GET';
    }

    var req = options.nativeProtocol.request(options, cb);

    // .... you might need to resend some data here for PUT, PUSH, etc. 

    if (res) {
      // We leave the user to call `end` on the first request
      req.end();
    }

    return req;
  }
};
```

This certainly is not convenient, but we are working on improving that.

**Comment by amitmangalvedkar** [comment]
Is there any code-sample for the 'undocumented' makeRequest. Although the above explanation does contain code, it would be really helpful if a complete working code could be made available. 
Also, is there any timeline by when the V1.0.0 is expected?

**Comment by jamestalmage** [comment]
Check out this [from the tests](https://github.com/olalonde/follow-redirects/blob/master/test/lib/https-config.js#L34).

It doesn't handle 307. To be perfectly honest, I don't know how to handle a 307 code correctly (simply because I haven't spent time researching it). The biggest issue is requests where you might be streaming a body **to** the server. How do we handle that? My guess is cache it and resend. How do we effectively handle caching for really large data? Is it typical for the redirecting server to consume the stream and send a redirect code? (i.e. is any of the body ever sent to the initial server), or is it more typical that the server sends the 307 early, and doesn't really consume the request body.

These are all pretty easy to figure out, but time consuming, and time is something I'm short on currently.

makeRequest is undocumented for a reason. It's not a great API, and we don't officially support it.

For now, you're only recourse may be to go back to the standard `http` and handle redirects manually. I will surely ping you once we get this issue fixed.

sorry I can't be more helpful in the short term.

**Comment by jamestalmage** [comment]
Closed by #28
